File: src/foundation/server/attention.rs
//! The attention lane — the web face reports its *own* window coming forward.
//!
//! `POST /api/in/attention` is a first-party heartbeat the page sends when it
//! becomes visible or regains focus (its `visibilitychange` / `focus` events).
//! It's the one signal that tells presence a window was *activated* (brought to
//! front), not merely left open — the strongest "they're checking on you" cue for
//! the expectation axis. Strictly first-party: the page reports only about its own
//! visibility, never anything about other apps or the wider system.
//!
//! Body-less and cheap; it just pokes [`crate::body::presence::Presence`].

use std::sync::Arc;

use axum::extract::State;
use axum::http::StatusCode;

use crate::foundation::server::headers::SceneHeader;
use crate::foundation::server::AppState;

/// Record that this scene's window was activated (became visible / focused).
pub async fn post_attention(
    State(state): State<Arc<AppState>>,
    SceneHeader(scene): SceneHeader,
) -> StatusCode {
    state.presence.note_activation(&scene);
    StatusCode::ACCEPTED
}

File: src/appearance/web/src/channels/in/attention.ts
// Client for the attention lane.
//
// `reportAttention` tells the backend this window just came forward — became
// visible or regained focus (POST /api/in/attention). It's the first-party
// "they're checking on you" signal for presence: strictly about *our own*
// window, never anything about other apps. Fire-and-forget; a failed beat is
// harmless and swallowed.

/** Report that our window was activated (became visible / focused). */
export async function reportAttention(opts: {
  scene: string;
  signal?: AbortSignal;
}): Promise<void> {
  try {
    await fetch("/api/in/attention", {
      method: "POST",
      headers: { "X-HI-Scene": opts.scene },
      signal: opts.signal,
    });
  } catch {
    /* a dropped heartbeat is harmless — presence just misses one activation */
  }
}
